// @sendbird/uikit-react/Channel/hooks/useHandleUploadFiles.js
import { b as __awaiter, c as __generator } from '../../chunks/bundle-Y6TqPszM.js';
import React__default, { useCallback } from 'react';
import { useSendbirdStateContext } from '../../useSendbirdStateContext.js';
import { f as isImage } from '../../chunks/bundle-ZT6XnGSN.js';
import { useGlobalModalContext } from '../../hooks/useModal.js';
import '../../chunks/bundle-0IelBspE.js';
import { ButtonTypes } from '../../ui/Button.js';
import { u as useLocalization } from '../../chunks/bundle-Xly_X4hP.js';
import { c as ModalFooter } from '../../chunks/bundle-5JvP07dI.js';
import { p as pxToNumber } from '../../chunks/bundle-2AUp3oeN.js';
import '../../withSendbird.js';
import '../../utils/message/getOutgoingMessageState.js';
import '../../chunks/bundle-zz1eCYds.js';
import '../../chunks/bundle-JhG5BHdE.js';
import 'react-dom';
import '../../chunks/bundle-JuKER7D_.js';
import '../../chunks/bundle-xgxXEKxk.js';
import '../../chunks/bundle-Mn36ivpf.js';
import '../../ui/IconButton.js';
import '../../ui/Icon.js';

var compressImage = function (_a) {
    var imageFile = _a.imageFile, compressionRate = _a.compressionRate, resizingWidth = _a.resizingWidth, resizingHeight = _a.resizingHeight;
    var image = document.createElement('img');
    return new Promise(function (resolve, reject) {
        image.src = URL.createObjectURL(imageFile);
        image.onerror = reject;
        image.onload = function () {
            URL.revokeObjectURL(image.src);
            var canvas = document.createElement('canvas');
            var originWidth = image.width;
            var originHeight = image.height;
            var widthRatio = originWidth / (resizingWidth || originWidth);
            var heightRatio = originHeight / (resizingHeight || originHeight);
            var targetResizingWidth = resizingWidth || originWidth;
            var targetResizingHeight = resizingHeight || originHeight;
            // Use the more impactful value, so the original images' ratio won't be broken
            if (widthRatio > heightRatio) {
                targetResizingHeight = originHeight / (resizingWidth ? widthRatio : 1);
            }
            else if (heightRatio > widthRatio) {
                targetResizingWidth = originWidth / (resizingHeight ? heightRatio : 1);
            }
            canvas.width = targetResizingWidth;
            canvas.height = targetResizingHeight;
            var ctx = canvas.getContext('2d');
            if (!ctx) {
                reject(new Error('Failed to get canvas 2d context'));
                return;
            }
            ctx.drawImage(image, 0, 0, targetResizingWidth, targetResizingHeight);
            ctx.canvas.toBlob(function (blob) {
                if (blob) {
                    var file = new File([blob], imageFile.name, { type: imageFile.type });
                    resolve(file);
                }
                else {
                    reject(new Error('Failed to compress image'));
                }
            }, imageFile.type, compressionRate);
        };
    });
};
var compressImages = function (_a) {
    var files = _a.files, logger = _a.logger, imageCompression = _a.imageCompression;
    return __awaiter(void 0, void 0, void 0, function () {
        var compressionRate, resizingWidth, resizingHeight, result;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    compressionRate = imageCompression.compressionRate;
                    resizingWidth = pxToNumber(imageCompression.resizingWidth);
                    resizingHeight = pxToNumber(imageCompression.resizingHeight);
                    result = {
                        failedIndexes: [],
                        compressedFiles: [],
                    };
                    if (!(Array.isArray(files) && files.length > 0)) {
                        logger.warning('utils - compressImages: There are no files.', files);
                        return [2 /*return*/, result];
                    }
                    if (compressionRate < 0 || 1 < compressionRate) {
                        logger.warning('utils - compressImages: The compressionRate is not acceptable.', compressionRate);
                        return [2 /*return*/, result];
                    }
                    return [4 /*yield*/, Promise.all(files
                            .map(function (file, index) { return __awaiter(void 0, void 0, void 0, function () {
                            var compressedImage, err_1;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        if (!(file.type === 'image/jpg' || file.type === 'image/png' || file.type === 'image/jpeg')) {
                                            logger.warning('utils - compressImages: The fileType is not compressible.', { file: file, index: index });
                                            result.failedIndexes.push(index);
                                            result.compressedFiles.push(file);
                                            return [2 /*return*/];
                                        }
                                        _a.label = 1;
                                    case 1:
                                        _a.trys.push([1, 3, , 4]);
                                        return [4 /*yield*/, compressImage({
                                                imageFile: file,
                                                compressionRate: compressionRate,
                                                resizingWidth: resizingWidth,
                                                resizingHeight: resizingHeight,
                                            })];
                                    case 2:
                                        compressedImage = _a.sent();
                                        result.compressedFiles.push(compressedImage);
                                        return [3 /*break*/, 4];
                                    case 3:
                                        err_1 = _a.sent();
                                        result.failedIndexes.push(index);
                                        logger.warning('utils - compressImages: Failed to compress image file', { err: err_1 });
                                        return [3 /*break*/, 4];
                                    case 4: return [2 /*return*/];
                                }
                            });
                        }); }))];
                case 1:
                    _b.sent();
                    logger.info('utils - compressImages: Finished compressing images', result);
                    return [2 /*return*/, result];
            }
        });
    });
};

var useHandleUploadFiles = function (_a, _b) {
    var sendFileMessage = _a.sendFileMessage, sendMultipleFilesMessage = _a.sendMultipleFilesMessage, quoteMessage = _a.quoteMessage;
    var logger = _b.logger;
    var stringSet = useLocalization().stringSet;
    var config = useSendbirdStateContext().config;
    var imageCompression = config.imageCompression;
    var uikitUploadSizeLimit = config === null || config === void 0 ? void 0 : config.uikitUploadSizeLimit;
    var uikitMultipleFilesMessageLimit = config === null || config === void 0 ? void 0 : config.uikitMultipleFilesMessageLimit;
    var openModal = useGlobalModalContext().openModal;
    var handleUploadFiles = useCallback(function (files) { return __awaiter(void 0, void 0, void 0, function () {
        var ONE_MiB, compressedFiles, sendingFiles, file, imageFiles_1, otherFiles_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    // Validate Parameters
                    if (!sendFileMessage || !sendMultipleFilesMessage) {
                        logger.warning('Channel|useHandleUploadFiles: required functions are undefined', { sendFileMessage: sendFileMessage, sendMultipleFilesMessage: sendMultipleFilesMessage });
                        return [2 /*return*/];
                    }
                    if (files.length === 0) {
                        logger.warning('Channel|useHandleUploadFiles: given file list is empty.', { files: files });
                        return [2 /*return*/];
                    }
                    if (files.length > uikitMultipleFilesMessageLimit) {
                        logger.info("Channel|useHandleUploadFiles: Cannot upload files more than ".concat(uikitMultipleFilesMessageLimit));
                        openModal({
                            modalProps: {
                                titleText: "Up to ".concat(uikitMultipleFilesMessageLimit, " files can be attached."),
                                hideFooter: true,
                            },
                            childElement: function (_a) {
                                var closeModal = _a.closeModal;
                                return (React__default.createElement(ModalFooter, { type: ButtonTypes.PRIMARY, submitText: stringSet.BUTTON__OK, hideCancelButton: true, onCancel: closeModal, onSubmit: closeModal }));
                            },
                        });
                        return [2 /*return*/];
                    }
                    // Validate file sizes
                    if (files.some(function (file) { return file.size > uikitUploadSizeLimit; })) {
                        // The default value of uikitUploadSizeLimit is 26MB
                        logger.info("Channel|useHandleUploadFiles: Cannot upload file size exceeding ".concat(uikitUploadSizeLimit));
                        ONE_MiB = 1024 * 1024;
                        openModal({
                            modalProps: {
                                titleText: "The maximum size per file is ".concat(Math.floor(uikitUploadSizeLimit / ONE_MiB), " MB."),
                                hideFooter: true,
                            },
                            childElement: function (_a) {
                                var closeModal = _a.closeModal;
                                return (React__default.createElement(ModalFooter, { type: ButtonTypes.PRIMARY, submitText: stringSet.BUTTON__OK, hideCancelButton: true, onCancel: closeModal, onSubmit: closeModal }));
                            },
                        });
                        return [2 /*return*/];
                    }
                    return [4 /*yield*/, compressImages({
                            files: files,
                            imageCompression: imageCompression,
                            logger: logger,
                        })];
                case 1:
                    compressedFiles = (_a.sent()).compressedFiles;
                    sendingFiles = compressedFiles;
                    // Send File Message
                    if (sendingFiles.length === 1) {
                        logger.info('Channel|useHandleUploadFiles: sending one file.');
                        file = sendingFiles[0];
                        sendFileMessage(file, quoteMessage);
                    }
                    else if (sendingFiles.length > 1) {
                        logger.info('Channel|useHandleUploadFiles: sending multiple files.');
                        imageFiles_1 = [];
                        otherFiles_1 = [];
                        sendingFiles.forEach(function (file) {
                            if (isImage(file.type)) {
                                imageFiles_1.push(file);
                            }
                            else {
                                otherFiles_1.push(file);
                            }
                        });
                        return [2 /*return*/, otherFiles_1.reduce(function (previousPromise, item) {
                                return previousPromise.then(function () {
                                    return sendFileMessage(item, quoteMessage);
                                });
                            }, (function () {
                                if (imageFiles_1.length === 0) {
                                    return Promise.resolve();
                                }
                                else if (imageFiles_1.length === 1) {
                                    return sendFileMessage(imageFiles_1[0], quoteMessage);
                                }
                                else {
                                    return sendMultipleFilesMessage(imageFiles_1, quoteMessage);
                                }
                            })())];
                    }
                    return [2 /*return*/];
            }
        });
    }); }, [
        sendFileMessage,
        sendMultipleFilesMessage,
        quoteMessage,
    ]);
    return handleUploadFiles;
};

export { useHandleUploadFiles };
