// @sendbird/uikit-tools/dist/cjs/uikitConfiguration/internal/initialConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initialRemoteConfig = exports.initialConfig = exports.createInitialRemoteConfig = exports.createInitialConfig = void 0;
const createInitialConfig = () => ({
    common: {
        enableUsingDefaultUserProfile: false,
    },
    groupChannel: {
        channel: {
            enableMention: false,
            enableOgtag: true,
            enableReactions: true,
            enableTypingIndicator: true,
            enableVoiceMessage: false,
            input: {
                camera: {
                    enablePhoto: true,
                    enableVideo: true,
                },
                enableDocument: true,
                gallery: {
                    enablePhoto: true,
                    enableVideo: true,
                },
            },
            replyType: 'quote_reply',
            threadReplySelectType: 'thread',
            // TODO: tw naming review
            typingIndicatorTypes: new Set(['text']),
        },
        channelList: {
            enableMessageReceiptStatus: false,
            enableTypingIndicator: false,
        },
        setting: {
            enableMessageSearch: false,
        },
    },
    openChannel: {
        channel: {
            enableOgtag: true,
            input: {
                camera: {
                    enablePhoto: true,
                    enableVideo: true,
                },
                enableDocument: true,
                gallery: {
                    enablePhoto: true,
                    enableVideo: true,
                },
            },
        },
    },
});
exports.createInitialConfig = createInitialConfig;
function deleteLocalConfigs(config) {
    // @ts-ignore
    delete config.groupChannel.channel.typingIndicatorTypes;
    return config;
}
const createInitialRemoteConfig = () => {
    const config = (0, exports.createInitialConfig)();
    return deleteLocalConfigs(config);
};
exports.createInitialRemoteConfig = createInitialRemoteConfig;
exports.initialConfig = (0, exports.createInitialConfig)();
exports.initialRemoteConfig = (0, exports.createInitialRemoteConfig)();
