// @sendbird/uikit-tools/dist/esm/uikitConfiguration/internal/initialConfig.js
export const createInitialConfig = () => ({
    common: {
        enableUsingDefaultUserProfile: false,
    },
    groupChannel: {
        channel: {
            enableMention: false,
            enableOgtag: true,
            enableReactions: true,
            enableTypingIndicator: true,
            enableVoiceMessage: false,
            input: {
                camera: {
                    enablePhoto: true,
                    enableVideo: true,
                },
                enableDocument: true,
                gallery: {
                    enablePhoto: true,
                    enableVideo: true,
                },
            },
            replyType: 'quote_reply',
            threadReplySelectType: 'thread',
            // TODO: tw naming review
            typingIndicatorTypes: new Set(['text']),
        },
        channelList: {
            enableMessageReceiptStatus: false,
            enableTypingIndicator: false,
        },
        setting: {
            enableMessageSearch: false,
        },
    },
    openChannel: {
        channel: {
            enableOgtag: true,
            input: {
                camera: {
                    enablePhoto: true,
                    enableVideo: true,
                },
                enableDocument: true,
                gallery: {
                    enablePhoto: true,
                    enableVideo: true,
                },
            },
        },
    },
});
function deleteLocalConfigs(config) {
    // @ts-ignore
    delete config.groupChannel.channel.typingIndicatorTypes;
    return config;
}
export const createInitialRemoteConfig = () => {
    const config = createInitialConfig();
    return deleteLocalConfigs(config);
};
export const initialConfig = createInitialConfig();
export const initialRemoteConfig = createInitialRemoteConfig();
