// @sendbird/uikit-tools/dist/cjs/hooks/useGroupChannelList/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useGroupChanelList = void 0;
const react_1 = require("react");
const groupChannel_1 = require("@sendbird/chat/groupChannel");
const premiumFeatures_js_1 = require("../../constants/premiumFeatures.js");
const useAsyncEffect_js_1 = require("../useAsyncEffect.js");
const useGroupChannelHandler_js_1 = require("../useGroupChannelHandler.js");
const usePreservedCallback_js_1 = require("../usePreservedCallback.js");
const reducer_js_1 = require("./reducer.js");
const createGroupChannelListCollection = (sdk, collectionCreator) => {
    const passedCollection = collectionCreator === null || collectionCreator === void 0 ? void 0 : collectionCreator();
    if (passedCollection)
        return passedCollection;
    const filter = new groupChannel_1.GroupChannelFilter();
    filter.includeEmpty = false;
    return sdk.groupChannel.createGroupChannelCollection({
        filter,
        limit: 20,
        order: groupChannel_1.GroupChannelListOrder.LATEST_LAST_MESSAGE,
    });
};
const useGroupChanelList = (sdk, options) => {
    var _a, _b;
    const deliveryReceiptEnabled = Boolean((_b = (_a = sdk.appInfo) === null || _a === void 0 ? void 0 : _a.premiumFeatureList) === null || _b === void 0 ? void 0 : _b.includes(premiumFeatures_js_1.premiumFeatures.delivery_receipt));
    const collectionRef = (0, react_1.useRef)();
    const { initialized, groupChannels, refreshing, appendChannels, deleteChannels, updateRefreshing, updateInitialized, } = (0, reducer_js_1.useGroupChannelListReducer)();
    const updateChannelsAndMarkAsDelivered = (shouldMarkAsDelivered, source, updatedChannels) => {
        var _a, _b, _c;
        const channels = (_b = (_a = collectionRef.current) === null || _a === void 0 ? void 0 : _a.channels) !== null && _b !== void 0 ? _b : [];
        appendChannels(channels, true);
        if (shouldMarkAsDelivered && deliveryReceiptEnabled) {
            switch (source) {
                case groupChannel_1.GroupChannelEventSource.EVENT_MESSAGE_RECEIVED:
                case groupChannel_1.GroupChannelEventSource.EVENT_MESSAGE_SENT:
                case groupChannel_1.GroupChannelEventSource.SYNC_CHANNEL_BACKGROUND:
                case groupChannel_1.GroupChannelEventSource.SYNC_CHANNEL_CHANGELOGS:
                case undefined:
                    (_c = options.markAsDelivered) === null || _c === void 0 ? void 0 : _c.call(options, updatedChannels !== null && updatedChannels !== void 0 ? updatedChannels : channels);
                    break;
            }
        }
    };
    const init = (0, usePreservedCallback_js_1.usePreservedCallback)((uid) => __awaiter(void 0, void 0, void 0, function* () {
        var _c, _d, _e, _f;
        if (collectionRef.current)
            (_c = collectionRef.current) === null || _c === void 0 ? void 0 : _c.dispose();
        if (uid) {
            collectionRef.current = createGroupChannelListCollection(sdk, options === null || options === void 0 ? void 0 : options.collectionCreator);
            (_d = collectionRef.current) === null || _d === void 0 ? void 0 : _d.setGroupChannelCollectionHandler({
                onChannelsAdded: (context, channels) => {
                    updateChannelsAndMarkAsDelivered(true, context.source, channels);
                },
                onChannelsUpdated: (context, channels) => {
                    updateChannelsAndMarkAsDelivered(true, context.source, channels);
                },
                onChannelsDeleted: () => {
                    updateChannelsAndMarkAsDelivered(false);
                },
            });
            if ((_e = collectionRef.current) === null || _e === void 0 ? void 0 : _e.hasMore) {
                yield ((_f = collectionRef.current) === null || _f === void 0 ? void 0 : _f.loadMore());
                updateChannelsAndMarkAsDelivered(true);
            }
        }
    }));
    (0, react_1.useEffect)(() => {
        return () => {
            var _a;
            if (collectionRef.current)
                (_a = collectionRef.current) === null || _a === void 0 ? void 0 : _a.dispose();
        };
    }, []);
    (0, useAsyncEffect_js_1.useAsyncEffect)(() => __awaiter(void 0, void 0, void 0, function* () {
        var _g;
        updateInitialized(false);
        yield init((_g = sdk.currentUser) === null || _g === void 0 ? void 0 : _g.userId);
        updateInitialized(true);
    }), [sdk, sdk.currentUser]);
    (0, useGroupChannelHandler_js_1.useGroupChannelHandler)(sdk, {
        onUserBanned: (channel, user) => {
            var _a;
            const isMe = user.userId === ((_a = sdk.currentUser) === null || _a === void 0 ? void 0 : _a.userId);
            if (isMe)
                deleteChannels([channel.url]);
            else
                updateChannelsAndMarkAsDelivered(false);
        },
    });
    const refresh = (0, usePreservedCallback_js_1.usePreservedCallback)(() => __awaiter(void 0, void 0, void 0, function* () {
        var _h;
        updateRefreshing(true);
        yield init((_h = sdk.currentUser) === null || _h === void 0 ? void 0 : _h.userId);
        updateRefreshing(false);
    }));
    const loadMore = (0, usePreservedCallback_js_1.usePreservedCallback)(() => __awaiter(void 0, void 0, void 0, function* () {
        var _j, _k;
        if ((_j = collectionRef.current) === null || _j === void 0 ? void 0 : _j.hasMore) {
            yield ((_k = collectionRef.current) === null || _k === void 0 ? void 0 : _k.loadMore());
            updateChannelsAndMarkAsDelivered(true);
        }
    }));
    return {
        /**
         * Initializing state, only available on first render
         * */
        initialized,
        /**
         * Get group channels state
         * */
        groupChannels,
        /**
         * Refresh, clear and reload messages from latest
         * @return {Promise<void>}
         * */
        refresh,
        /**
         * Refreshing state, status is changes when the refresh is called.
         * */
        refreshing,
        /**
         * Fetch more channels to state
         * @return {Promise<void>}
         * */
        loadMore,
    };
};
exports.useGroupChanelList = useGroupChanelList;
